Introduce `masterServiceAnnotations` & `replicaServiceAnnotations`

owenthereal opened the pull request:
Introduce `masterServiceAnnotations` & `replicaServiceAnnotations` to the `Postgresql` CRD. The following behavior is implemented

* `masterServiceAnnotations` takes precedence over `serviceAnnotations` for master service if not empty.
* `replicaServiceAnnotations` takes precedence over `serviceAnnotations` for replica service if not empty.
* The existing definition of `serviceAnnotations` continue to work for backward compatibility when neither `masterServiceAnnotations` nor `replicaServiceAnnotations` are defined.

This closes https://github.com/zalando/postgres-operator/issues/1927

FxKu commented on docs/reference/cluster_manifest.md:
```diff
@@ -173,6 +173,20 @@ These parameters are grouped directly under  the `spec` key in the manifest.
+  to the master services created for the database cluster. Check the
+  [administrator docs](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#load-balancers-and-allowed-ip-ranges)
+  for more information regarding default values and overwrite rules.
+  This field takes precedence over `serviceAnnotations` for the master services if not empty.
```
I guess singular is better: "to the master service of ..." (same for replica service)

owenthereal replied:
Addressed in https://github.com/zalando/postgres-operator/pull/2161/commits/eea64b600cf1d2f09262772a69ce33ef4674119b

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -1905,12 +1905,33 @@ func (c *Cluster) generateServiceAnnotations(role PostgresRole, spec *acidv1.Pos
+				specServiceAnnotations = spec.ServiceAnnotations
+			}
+		default:
+			specServiceAnnotations = spec.ServiceAnnotations
```
This is either master|replica service annotation OR service annotation, right?
It should be accumulative. First, global config, then ServiceAnnotations overriding, then MasterServiceAnnotations and ReplicaServiceAnnotations. (maybe there's a golang function so we don't need 4 loops)

owenthereal replied:
Addressed in https://github.com/zalando/postgres-operator/pull/2161/commits/03878300fb17c5c46e7b621551ba8b249a4aca2e

FxKu commented on pkg/cluster/cluster_test.go:
```diff
@@ -792,21 +811,38 @@ func TestServiceAnnotations(t *testing.T) {
+				"external-dns.alpha.kubernetes.io/hostname":                            "acid-test-repl.test.db.example.com,test-repl.acid.db.example.com",
+				"service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout": "2000",
+			},
+		},
```
there could also be a test to check if annotations from `serviceAnnotations` are taken if they do not exist in master|replicaServiceAnnotations (could also be included here I think)

owenthereal replied:
Addressed in https://github.com/zalando/postgres-operator/pull/2161/commits/03878300fb17c5c46e7b621551ba8b249a4aca2e

FxKu commented:
Could you add the new field also to the CRD schema (please, follow alphabetical sorting)
- [example manifest](https://github.com/zalando/postgres-operator/blob/master/manifests/postgresql.crd.yaml)
- [helm chart](https://github.com/zalando/postgres-operator/blob/master/charts/postgres-operator/crds/postgresqls.yaml)
- [go representation](https://github.com/zalando/postgres-operator/blob/master/pkg/apis/acid.zalan.do/v1/crds.go#L137)

owenthereal commented:
@FxKu:


> Could you add the new field also to the CRD schema (please, follow alphabetical sorting)
>
> [example manifest](https://github.com/zalando/postgres-operator/blob/master/manifests/postgresql.crd.yaml)
> [helm chart](https://github.com/zalando/postgres-operator/blob/master/charts/postgres-operator/crds/postgresqls.yaml)
> [go representation](https://github.com/zalando/postgres-operator/blob/master/pkg/apis/acid.zalan.do/v1/crds.go#L137)

Addressed in https://github.com/zalando/postgres-operator/pull/2161/commits/030c482cfac888c1da9d136c6bad3a2b05aacfce & https://github.com/zalando/postgres-operator/pull/2161/commits/db7befa982c613f01e6e4bbd38d149cc1c83ad15

Mind taking another look? I have addressed all feedback, and this PR should be ready to merge. Cheers!

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -1905,9 +1905,15 @@ func (c *Cluster) generateServiceAnnotations(role PostgresRole, spec *acidv1.Pos
 	for k, v := range c.OpConfig.CustomServiceAnnotations {
 		annotations[k] = v
 	}
```
```suggestion
	annotations = mergeAnnotations(annotations, c.OpConfig.CustomServiceAnnotations)
```
:smiley:

FxKu commented on docs/reference/cluster_manifest.md:
```diff
@@ -173,6 +173,20 @@ These parameters are grouped directly under  the `spec` key in the manifest.
 
+* **masterServiceAnnotations**
+  A map of key value pairs that gets attached as [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)
+  to the master services created for the database cluster. Check the
```
```suggestion
  to the master service created for the database cluster. Check the
```
same for replica

FxKu commented on docs/reference/cluster_manifest.md:
```diff
@@ -173,6 +173,20 @@ These parameters are grouped directly under  the `spec` key in the manifest.
+  to the master services created for the database cluster. Check the
+  [administrator docs](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#load-balancers-and-allowed-ip-ranges)
+  for more information regarding default values and overwrite rules.
+  This field overrides `serviceAnnotations` for the master service if not empty.
```
```suggestion
  This field overrides `serviceAnnotations` with the same key for the master
  service if not empty.
```

owenthereal replied:
I updated the admin doc here: https://github.com/zalando/postgres-operator/pull/2161/commits/59016029914b694bb66b1e9db72de2bcc67f42be

dmvolod commented on pkg/generated/clientset/versioned/fake/register.go:
```diff
@@ -45,14 +45,14 @@ var localSchemeBuilder = runtime.SchemeBuilder{
-//     clientsetscheme "k8s.io/client-go/kubernetes/scheme"
-//     aggregatorclientsetscheme "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/scheme"
-//   )
+//	import (
+//	  "k8s.io/client-go/kubernetes"
+//	  clientsetscheme "k8s.io/client-go/kubernetes/scheme"
+//	  aggregatorclientsetscheme "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/scheme"
+//	)
```
@owenthereal what is your golang version installed. For 1.17 and 1.18 generated files should not be changed

owenthereal replied:
I'm using 1.19 locally. I regenerated the code with 1.18 in https://github.com/zalando/postgres-operator/pull/2161/commits/3fb0b42f27c025ffe9e8088088ab0007262264f8. 

It would be nice to introduce https://asdf-vm.com or the like to this project to auto-switch or prompt to install the right Go version.

dmvolod replied:
Not sure we need an additional tools for that, as golang supports multiple versions management by itself
https://go.dev/doc/manage-install

As an option, some validation steps with autogenrration/formatting/linting can be added as part of the #2135

dmvolod commented on pkg/cluster/k8sres.go:
```diff
@@ -2380,3 +2384,18 @@ func ensurePath(file string, defaultDir string, defaultFile string) string {
 }
+
+// mergeAnnotations merged annotations that entries in b override the ones in a.
+func mergeAnnotations(a, b map[string]string) map[string]string {
```
When we will be on golang 1.18 with #2164, `maps.Copy` from the `golang.org/x/exp/maps` can be used instead.
Seems to this is too much resource and memory consumption implementation.
@FxKu wdyt about it?

FxKu replied:
I read about `maps.Copy` yesterday. We would not need this function then and could do it in the code. Sounds good to me.

owenthereal replied:
Good call! Switched to use `maps.Copy` in https://github.com/zalando/postgres-operator/pull/2161/commits/3fb0b42f27c025ffe9e8088088ab0007262264f8

FxKu commented:
:+1:

idanovinda commented:
👍
